File: pkg/heap/top_k_frequent.go
package heap

import "container/heap"

type Pair struct {
	Val  int
	Freq int
}

/*
 * 347. Top K Frequent Elements
 */

func topKFrequent(nums []int, k int) []int {
	freqMap := make(map[int]int)
	for _, v := range nums {
		freqMap[v] += 1
	}

	maxHeap := &Heap[Pair]{
		less: func(i Pair, j Pair) bool {
			return i.Freq > j.Freq
		},
	}
	for k, v := range freqMap {
		heap.Push(maxHeap, Pair{
			Val:  k,
			Freq: v,
		})
	}

	res := make([]int, k)
	for i := 0; i < k; i++ {
		item := heap.Pop(maxHeap).(Pair)
		res[i] = item.Val
	}
	return res
}

File: python/heap/top_k_frequent.py
from collections import Counter
from typing import List
import heapq

"""
347. Top K Frequent Elements
"""
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = Counter(nums)

        heap: List[tuple[int, int]] = []
        for num, freq in count.items():
            heapq.heappush(heap, (freq, num))
            if len(heap) > k:
                heapq.heappop(heap)   # evict least frequent

        return [num for _, num in heap]
    
    def topKFrequentClaude(self, nums: List[int], k: int) -> List[int]:
        count = Counter(nums)
        return heapq.nlargest(k, count.keys(), key=lambda x: count[x])